lib/libmbedtls/core/tee_lmd_provider.c

```diff
@@ -554,92 +642,567 @@ void crypto_bignum_clear(struct bignum *s)
+					   uint8_t *dst, size_t *dst_len)
+{
+	TEE_Result res = TEE_SUCCESS;
+	mbedtls_rsa_context rsa;
+	int lmd_res;
+	uint8_t *buf = NULL;
+	unsigned long blen, offset;
+
+	rsa_init_from_key_pair(&rsa, key);
+
+	blen = CFG_CORE_BIGNUM_MAX_BITS / 8;
+	buf = malloc(blen);
+	if (!buf) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	memset(buf, 0, blen);
+	memcpy(buf + rsa.len - src_len, src, src_len);
+
+	lmd_res = mbedtls_rsa_private(&rsa, NULL, NULL, buf, buf);
+	if (lmd_res != 0) {
+		FMSG("mbedtls_rsa_private() returned 0x%x", -lmd_res);
+		res = get_tee_result(lmd_res);
+		goto out;
+	}
+
+	/* Remove the zero-padding (leave one zero if buff is all zeroes) */
+	offset = 0;
+	while ((offset < rsa.len - 1) && (buf[offset] == 0))
+		offset++;
+
+	*dst_len = rsa.len - offset;
+	memcpy(dst, (char *)buf + offset, *dst_len);
+out:
+	if (buf)
+		free(buf);
+	rsa_free(&rsa);
+	return res;
 }
 
-TEE_Result crypto_acipher_rsaes_decrypt(uint32_t algo __unused,
-					struct rsa_keypair *key __unused,
+TEE_Result crypto_acipher_rsaes_decrypt(uint32_t algo, struct rsa_keypair *key,
 					const uint8_t *label __unused,
 					size_t label_len __unused,
-					const uint8_t *src __unused,
-					size_t src_len __unused,
-					uint8_t *dst __unused,
-					size_t *dst_len __unused)
+					const uint8_t *src, size_t src_len,
+					uint8_t *dst, size_t *dst_len)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	TEE_Result res = TEE_SUCCESS;
+	int lmd_res, lmd_padding;
+	size_t blen;
+	size_t mod_size;
+	void *buf = NULL;
+	mbedtls_rsa_context rsa;
+	const mbedtls_pk_info_t *pk_info = NULL;
+	uint32_t md_algo = MBEDTLS_MD_NONE;
+
+	rsa_init_from_key_pair(&rsa, key);
+
+	/*
+	 * Use a temporary buffer since we don't know exactly how large
+	 * the required size of the out buffer without doing a partial
+	 * decrypt. We know the upper bound though.
+	 */
+	if (algo == TEE_ALG_RSAES_PKCS1_V1_5) {
+		mod_size = crypto_bignum_num_bytes(key->n);
+		blen = mod_size - 11;
+		lmd_padding = MBEDTLS_RSA_PKCS_V15;
+	} else {
+		/* Decoded message is always shorter than encrypted message */
+		blen = src_len;
+		lmd_padding = MBEDTLS_RSA_PKCS_V21;
+	}
+
+	buf = malloc(blen);
+	if (!buf) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	pk_info = mbedtls_pk_info_from_type(MBEDTLS_PK_RSA);
+	if (pk_info == NULL) {
+		res = TEE_ERROR_NOT_SUPPORTED;
+		goto out;
+	}
+
+	/*
+	 * TEE_ALG_RSAES_PKCS1_V1_5 is invalid in hash. But its hash algo will
+	 * not be used in rsa, so skip it here.
+	 */
+	if (algo != TEE_ALG_RSAES_PKCS1_V1_5) {
+		md_algo = tee_algo_to_mbedtls_hash_algo(algo);
+		if (md_algo == MBEDTLS_MD_NONE) {
+			res = TEE_ERROR_NOT_SUPPORTED;
+			goto out;
+		}
+	}
+
+	mbedtls_rsa_set_padding(&rsa, lmd_padding, md_algo);
+
+	if (lmd_padding == MBEDTLS_RSA_PKCS_V15)
+		lmd_res = pk_info->decrypt_func(&rsa, src, src_len, buf, &blen,
+						blen, NULL, NULL);
+	else
+		lmd_res = pk_info->decrypt_func(&rsa, src, src_len, buf, &blen,
+						blen, mbd_rand, NULL);
```

Comment:
Would it be an error to always supply `mbd_rand()`?
If not then we could simplify here.

Reply:
No, for `MBEDTLS_RSA_PKCS_V15`, it must not to supply rand() function, for it will do some blinding operation (for DQ, DP ...) with rand(). But in DQ and DP may be not supplied by the caller, so it will get failed. 
But for no V15 algo, we must add rand() function.

Reply:
OK